### ISSUE TITLE
Multi-turn moves are visible to enemy players

### ISSUE THREAD

**Issue by stevecotton** [wesnoth/wesnoth#6292]
### Game and System Information
- reproduced on 1.16.1+dev (e74ceb44329d416f547b17087d432522b27d2a07), Linux local build
- originally reported on 1.16.1(as tagged), by ben24626 during [a ladder match against Checkhovs_Cannon](https://replays.wesnoth.org/1.16/2021/11/12/2p__Caves_of_the_Basilisk_Turn_22_(2678).bz2)
- Note: ladder-match replays are broken by #6283.

### Describe the bug
In some circumstances, preset moves for future turns are visible to opponents and observers.

In trying to reproduce it, it seems the unit must move at least the first hex of a multi-turn move to trigger the bug. If the unit already has no movement points remaining, then setting a preset move for the next turn doesn't leak to the other players, even if it's a multi-turn move.

In the replay of Ben's game, we see this move by Checkhov, followed by Ben asking "do you have a preset move for 13,16 to move to 10,17?" "cause the computer is showing me that"
```
	[command]
		from_side=2
		sent=yes
		[move]
			x="17,16,15,14,13,12,11,10"
			y="15,15,16,15,16,16,17,17"
		[/move]
		[checkup]
			[result]
				final_hex_x=13
				final_hex_y=16
				stopped_early=no
			[/result]
			[result]
				next_unit_id=58
				random_calls=0
			[/result]
		[/checkup]
	[/command]
```

Screenshot from an attempt to reproduce it (with different units):
![opponent_preset_move](https://user-images.githubusercontent.com/101462/141669692-2927dcae-6a70-47f5-bb27-09c92ee3c75d.png)

**Comment by Wedge009** [comment]
I sort-of confirmed this via local MP in Caves of Basilisk. Just recruit scouting units on each side and send them to the other team's camp to get into visible range. Doesn't look like anything is amiss in 1.14.9 (what I have available in distribution repository), but enemy movements occasionally appeared in 1.16.1 official release build.

As per the description, it seems to happen when in the middle of a move (halted upon 'enemy sighted' event). It doesn't happen consistently, and after a while it looks like it's replace with the normal 'reach map' that's displayed upon hovering the cursor over a unit, whether allied or enemy. The inconsistency of replicating the symptoms will make it tricky to resolve, I think.

**Comment by Wedge009** [comment]
Interesting - this doesn't look like it's specific to MP. I was playing Hide and Seek from Liberty (enemy infantry are all static unless you enter their sight range). In the screenshot the enemy infantry is targeting my location _before_ the move, which I'm guessing the AI would take action _if_ they were able to see me. I suspect this revelation wouldn't be showing in 1.14.x and also I cannot consistently replicate it...

![hideandseek](https://user-images.githubusercontent.com/13679322/141732068-9ba2f88f-949c-4c03-9203-6f8295ced912.png)

**Comment by stevecotton** [comment]
1.14.x certainly had it visible to allies, although I only remember that in SXC (cooperative) when someone was lagging/disconnected, and we noticed at the start of their turn that they had a preset move. That we saw it multiple times there was probably due to SXC's play style - long matches, along with times when someone would need to return to a shop in a known-safe location.

**Comment by Wedge009** [comment]
Allies makes sense. Enemies, no. I thought the latter was what this report was about.

**Comment by Wedge009** [comment]
I think there is something related to the 'enemy sighted!' event that triggers this issue. My recollection is that I have not encountered the problem where there is no shroud or fog.

**Comment by Wedge009** [comment]
I find I can replicate it fairly consistently when I get the 'enemy sighted, press t to continue' event/message.

**Comment by stevecotton** [comment]
@Wedge009 are you only seeing it during the enemy player's turn? I've found some code that's checking whose turn it is, when it should be checking whose GUI it's drawing; the question is to check whether I've found this bug or a different bug.

**Comment by Wedge009** [comment]
I can see enemy's planned moves when it's my turn. I'm not explicitly hunting for it, but I've noticed it come up a few times as I've been playing through (SP) campaigns in 1.16.1 release. I generally don't make multi-turn moves myself and I wouldn't be looking at my own moves when it's the enemy's turn anyway.

**Comment by stevecotton** [comment]
When I next update that WIP, I'll note that it fixes something but not the whole bug then :(

**Comment by Wedge009** [comment]
I didn't test your commit specifically, just remarking what I've been observing during SP campaigns. Or are you saying what you have so far only deals with MP situations specifically?

**Comment by stevecotton** [comment]
My WIP deals with both SP and MP, but the bug that it fixes can only happen during the enemy turn. If you've seen something similar during your own turn then we've a second bug to find.

**Comment by Wedge009** [comment]
Hmm, in SP at least, I don't recall this occurring during the enemy's turn (as mentioned above it's easiest to replicate when playing with shroud/fog and triggering an 'enemy sighted' event). And 'enemy' in MP is relative.

**Comment by stevecotton** [comment]
When triggering an 'enemy sighted' event, are you seeing an enemy unit's plans, or are you seeing your own unit's plans?

**Comment by Wedge009** [comment]
Playing in SP, exposing shroud/fog, finding an enemy unit -> not sure of the exact circumstances to trigger it but often when hovering over enemy units I can see their planned moves. As you just mentioned, there is the Hide and Seek example I gave above.

Not sure what effect the different AI sides would have (there is one for the enemy established at a keep, one for the infantry scattered around, and one for the rogues that Baldras is meeting), but yes I've been finding this issue in many SP scenarios. I cannot recall seeing this in 1.14 - and I was playing through several campaigns in 1.14.17 just before 1.16.0 released.

**Comment by stevecotton** [comment]
Okay, I think I've found that one. If it's what I think it is, mousing over any other unit (including your own units) shows the path to the same hex. If you bind a button to the 'select hex' hotkey then you'll see the effect as it's meant to be seen, with some highlights on top.

If it's that, then it's caused by the `// Show the route from selected hex to mouseover unit` code in mouse_event.cpp (and yes, the comment says from selected hex to unit, but that's backwards). I think it's not leaking enemy plans, rather it's a hex chosen based on local UI input, combined with a bug that the hex should have been deselected.

**Comment by Wedge009** [comment]
Hmm, I'll try to have a look more closely the next time I encounter this issue. But for the Hide and Seek example I gave above it did seem to be showing the enemy's plans because it was targeting my units - or at least where they were at the start of the turn. I haven't used this 'select hex' function before and it's also odd that I haven't seen it in 1.14.

**Comment by stevecotton** [comment]
I think the hex got selected by a UI event (so tending to be where you clicked on your unit to start moving it), and there's a not-in-1.14 bug that that selected hex didn't get cleared by the enemy sighted / continue move sequence.

**Comment by Wedge009** [comment]
Hmm, I'll try to check the next time I find the issue, but testing the 'select hex' option it looks like all units show the same multi-move path to the selected hex, not just a particular enemy unit. Or are you saying this hex selection can be assigned to a particular unit?

**Comment by Wedge009** [comment]
Okay, I managed to replicate the situation you described - getting past the 'enemy sighted' event, it seems to trigger a 'select hex' action on my moving unit and all the enemy units are pointing to that 'selected' hex. So maybe I was getting that issue rather than the original MP one that you reported.

I'm not able to replicate in 1.14, though...

**Comment by stevecotton** [comment]
@Wedge009 do you have any more feedback?

I just realised that I haven't yet opened a PR for the fix that I have, so am intending to get that done today/tomorrow.

**Comment by Wedge009** [comment]
Only that it does indeed appear that the situation I'm seeing - while I don't remember seeing it in 1.14 - seems to be related to the 'select hex' trigger you described. When an 'enemy sighted' event is triggered, it looks like a 'select hex' is also triggered where my unit is and then every other unit I hover over - including the enemy ones - look like they're targeting that selected hex.

**Comment by Wedge009** [comment]
Just noting that I observed the 'select hex' issue again while doing unrelated testing in 1.17.2+dev. So it was a separate issue which is what I think we established previously anyway.

**Comment by stevecotton** [comment]
Any idea what triggered it?

Do you have a hotkey bound to selecting a hex?
Do you have #5892 merged into your dev build (I don't think it would cause this, but want to rule it out as a cause)?

**Comment by Wedge009** [comment]
No hot-key. I leave everything as default except for End Turn. It was a clean build off master (25e763f29be) on Linux.

I was just doing some debug-testing in NR Pursuit, relating to the spider den trap in the north-west. Removed shroud, debug-killed the orange-side leader which spawns the allied dwarves. I noticed after that that the dwarves and the remaining orange-side units were all highlighting paths to where the orange-side leader was killed.

I don't think this behaviour happens consistently though, so I imagine if you tried to replicate that, it probably won't happen for you.
